### PR TITLE
Add service for notification email and reset password

### PR DESCRIPTION
**Pull request by ImMin5**
### Category
- [x] New feature
- [ ] Bug fix
- [x] Improvement
- [ ] Refactor
- [ ] etc

### Description
 - add email send service
 - add verify_email and confirm_email 
 - modify create 
 - add email template
 
### Known issue
-  #31

### CODE REVIEW

File: src/spaceone/identity/conf/global_conf.py [hunk 1]
```diff
@@ -24,6 +24,7 @@
 
 IDENTITY = {
     'token': {
+        'temporary_token_timeout': 3600,
```
Comment (whdalsrnt) [inline src/spaceone/identity/conf/global_conf.py:27]: change key of token configurations
`temporary_token_timeout` -> `verify_code_timeout`

File: src/spaceone/identity/error/error_user.py [hunk 2]
```diff
@@ -31,3 +31,23 @@ class ERROR_INCORRECT_USER_ID_FORMAT(ERROR_INVALID_ARGUMENT):
...
+
+class ERROR_INVALID_VERIFY_CODE(ERROR_INVALID_ARGUMENT):
+    _message = 'Invalid verify token. (verify_code = {verify_code})'
+
+
+class ERROR_UNABLE_TO_RESET_PASSWORD(ERROR_INVALID_ARGUMENT):
+    _message = 'Unable to reset password. (user_id = {user_id})'
+
+
+class ERROR_PASSWORD_NOT_CHANGED(ERROR_UNABLE_TO_RESET_PASSWORD):
```
Comment (whdalsrnt) [inline src/spaceone/identity/error/error_user.py:44]: change the message to
`The requested password is the same as the previous password.`

File: src/spaceone/identity/manager/email_manager.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class EmailManager(BaseManager):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.smtp_connector: SMTPConnector = self.locator.get_connector('SMTPConnector')
+
+    def send_reset_password_email(self, user_id, email, reset_password_link, language):
+        if not os.path.exists(f'{TEMPLATE_PATH}/reset_password_link_{language}.html'):
```
Comment (whdalsrnt) [inline src/spaceone/identity/manager/email_manager.py:25]: Remove template validation logic

File: src/spaceone/identity/manager/token_manager/__init__.py [hunk 4]
```diff
@@ -41,6 +46,7 @@ def _load_conf(self):
         identity_conf = config.get_global('IDENTITY') or {}
         token_conf = identity_conf.get('token', {})
         self.CONST_TOKEN_TIMEOUT = token_conf.get('token_timeout', 1800)
+        self.CONST_TEMPORARY_TOKEN_TIMEOUT = token_conf.get('temporary_token_timeout')
```
Comment (whdalsrnt) [inline src/spaceone/identity/manager/token_manager/__init__.py:49]: change variable name
`COST_TEMPROARY_TOKEN_TIMEOUT` -> `CONST_VERIFY_TOKEN_TIMEOUT`

File: src/spaceone/identity/manager/token_manager/__init__.py [hunk 5]
```diff
@@ -151,3 +160,18 @@ def _set_refresh_token_cache(self, new_refresh_key):
                 cache.delete(f'refresh-token:{self.old_refresh_key}')
 
             cache.set(f'refresh-token:{new_refresh_key}', '', expire=self.CONST_REFRESH_TIMEOUT)
+
+    def generate_verify_code(self, domain_id, user_id):
+        if cache.is_set():
+            verify_code = str(random.randint(100000, 999999))
```
Comment (whdalsrnt) [inline src/spaceone/identity/manager/token_manager/__init__.py:166]: separate the function that generates the verify code

**Review by whdalsrnt — CHANGES_REQUESTED** [review]
Hi, @ImMin5 

I reviewed the code you PR.
Please check the comments and change the code.

**Comment by ImMin5** [comment]
Hi, @whdalsrnt 

Thank you for your review. 
I have applied your feedback and made the necessary changes. 
Would you please take a look at the updated version and let me know if there is anything else I can do to improve it.

File: src/spaceone/identity/connector/smtp_connector.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        password = self.config.get('password')
+        self.from_email = self.config.get('from_email')
+        self.set_smtp(host, port, user, password)
+
+    def set_smtp(self, host, port, user, password):
+        self.smtp = smtplib.SMTP(host, port)
+        self.smtp.connect(host, port)
+        self.smtp.ehlo()
+        self.smtp.starttls()
+        self.smtp.ehlo()
```
Comment (whdalsrnt) [inline src/spaceone/identity/connector/smtp_connector.py:30]: Please remove duplicate code. 
`self.smtp.ehlo()`

File: src/spaceone/identity/manager/email_manager.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        subject = '[SpaceONE] Reset your password'
+
+        self.smtp_connector.send_email(email, subject, email_contents)
+
+    def send_verification_email(self, user_id, email, verification_code, language):
+        language = LANGUAGE_MAPPER.get(language, 'en')
+
+        template = JINJA_ENV.get_template(f'verification_code_{language}.html')
+        email_contents = template.render(user_name=user_id, verification_code=verification_code)
+        subject = '[SpaceONE] Verify your email address'
```
Comment (whdalsrnt) [inline src/spaceone/identity/manager/email_manager.py:42]: Please change the email subject and contents to be configurable.
- Location: `global_conf.py`
- Default: `Cloudforet`

File: src/spaceone/identity/manager/token_manager/__init__.py [hunk 8]
```diff
@@ -151,3 +160,22 @@ def _set_refresh_token_cache(self, new_refresh_key):
                 cache.delete(f'refresh-token:{self.old_refresh_key}')
 
             cache.set(f'refresh-token:{new_refresh_key}', '', expire=self.CONST_REFRESH_TIMEOUT)
+
+    def generate_verify_code(self, domain_id, user_id):
```
Comment (whdalsrnt) [inline src/spaceone/identity/manager/token_manager/__init__.py:164]: Duplicate function name below.
Please change it to `create_verify_code`.

**Review by whdalsrnt — CHANGES_REQUESTED** [review]

**Comment by ImMin5** [comment]
Hi, @whdalsrnt 
Thanks for spend time for review, I applied what you suggested.

**Review by whdalsrnt — APPROVED** [review]
Hi, @ImMin5 
Thanks for the quick change to the feedback.

**Pull request merged**